### ISSUE TITLE
Make selecting individual units more forgiving

### ISSUE THREAD

**Issue by GoogleFrog** [beyond-all-reason/spring#671]
Selecting individual units is needlessly hard in default spring. To reproduce.
1.  Press the mouse on point A.
2. Move the mouse to point B.
3.  Release the mouse.
![image](https://user-images.githubusercontent.com/1318249/222472133-8bc1ee51-a4e5-4171-8674-d8334ac77cda.png)

Step 2 draws a tiny selection box. Step 3 selects nothing because no unit midpoint is inside the box. Any player who does these steps expected to select the unit, but instead ends up with nothing.

This issue was raised by players just before ZK steam so I wrote this abomination of a widget: https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/gui_selection_modkeys.lua

I haven't had a problem with it in years, but it is the sort of thing that should be removed now that the engine dev exists to do so. It's basic behaviour is as follows.
- Remember the unitID underneath every mouse press.
- Detect whether a selection box was drawn.
- If a selection box was drawn, and no unit was selected, and the same unitID is under the mouse when released, select the unit.

It also does something to do with double clicks, or that might just be to better target its hax. Ideally the engine could do the single unit selection fix in a way that everyone likes and never has to think about.

I also made a similar widget to help with issuing point-commands on units (guard and area attack) since dragging those commands causes circles to appear by default. It is similarly bad, but less hax is involved to work around the selection system, and choices around area commands strays more into game design than the selection widget: https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/cmd_attack_command_helper.lua

**Comment by badosu** [comment]
I need to read a bit more about the problem and reproduce it.

I understand the rationale and proposed fix, one question: have you considered tweaking `MouseDragSelectionThreshold`?

If yes and it's insufficient for you, would a setting that considers a minimum world xz distance threshold for selection resolve this for you? (seems like one fix that resolves the issue for the particular zoomed in case and a natural improvement in general).

**Comment by badosu** [comment]
I'm working on a PoC for world distance threshold as opposed to screen pixels. I think it might resolve this problem more elegantly but I might be wrong.

**Comment by sprunk** [comment]
Doesn't a selection box fundamentally work on screen pixels? The problem is that a single point-click works on the hitvolume while an area-drag works only on the midpoint (I think, or generally some singular point).
![image](https://user-images.githubusercontent.com/2573076/222907499-cb766c30-dfbf-4844-8c44-fe0723305a27.png)

**Comment by badosu** [comment]
When threshold is lower then it would perform a trace which should be identical to click on pivot point (mouserelease).

So if for example your minimum threshold for selection in world coordinates is about the size of the smallest unit you have you'd ensure all units would be selectable on higher zoom. The selectbox would actually not event start, instead deferring to tracing a ray to the unit below mouse cursor.

**Comment by badosu** [comment]
I'd have to check with tracescreenray though what is the collision it performs to select a unit, but I suspect this would work. I still need to work on it, I have no working debug build at the moment.

**Comment by sprunk** [comment]
Sounds decent, though as far as I understand it it would very quickly diminish in usefulness as unit size discrepancy grows.

Consider these:
![image](https://user-images.githubusercontent.com/2573076/222924419-cd2fb90b-300a-4900-a31c-0a8b227bdb7d.png)
![image](https://user-images.githubusercontent.com/2573076/222924430-d0d8a728-0f0c-4573-9f0a-ea2f9c9ce4a8.png)

On both those screens there's two fleas, and a cyan selection box in the larger unit large enough to select those two fleas. I would want the threshold to be lower than the size of that box (so that it doesn't devolve into a point-click, because then you can't select 2 fleas). But it would fail to select the building in both cases.

**Comment by badosu** [comment]
What if you make minimum distance about 1 flea hitbox diameter.

**Comment by sprunk** [comment]
Why don't we want to trace ray on a big box?

**Comment by dzosz** [comment]
wouldn't adding single click behavior to box selection just work out of the box for this case?

**Comment by badosu** [comment]
> Why don't we want to trace ray on a big box?

Sounds to me like something that would be confusing. As a player I wouldn't expect to select a unit on edge of mouserelease just because the box was empty.

Perhaps being able to scan hitboxes on edges would do the job.

**Comment by sprunk** [comment]
> I wouldn't expect to select a unit on edge of mouserelease just because the box was empty.

But in that case it wasn't empty, there was a unit in the box after all, just its midpoint wasn't but that's a technicality.

> being able to scan hitboxes on edges would do the job.

Yes, this would be ideal. In general, it would be ideal if selection didn't work on unit midpoints but on hitvolumes, but under the assumption that the midpos is contained inside a hitvolume and that hitvolumes are convex this is equivalent to scanning just the edges.

**Comment by badosu** [comment]
> Yes, this would be ideal. In general, it would be ideal if selection didn't work on unit midpoints but on hitvolumes, but under the assumption that the midpos is contained inside a hitvolume and that hitvolumes are convex this is equivalent to scanning just the edges.

Yeah I think scanning hitbox on vertices would be ideal. In this case I need to check for a cheap method to do that (if there is one). Given the geometry it might be possible to do that without extreme computational cost.

**Comment by GoogleFrog** [comment]
I am not sure why we would want any selection to happen in world space. I haven't seen a reason to not just transfer my widget to the engine and call it done. We already have selection volumes, so I can't think of anything more I would want from point selection.

**Comment by badosu** [comment]
Here's a very crude implementation of intersection of selection rectangle with units iconRadius.

Lacks proper colvol handling for units and transformation matrix for the projection, but it kinda works when you look top-down: https://github.com/beyond-all-reason/spring/commit/1696974d7339aa320946602ba00152acf20337b9.

Next steps are handling colvol properly (get a true hitbox radius if sphere, test other types of colvol if necessary) and a proper projection for the camera.

**Comment by badosu** [comment]
If all units colvols are spheres then the change is not that expensive. Even if we increase the cost of selection we can gain by optimizing with quadrants (instead of iterating on all units).

**Comment by GoogleFrog** [comment]
I want units to be points for the purpose of including them in typical selection rectangles. I don't want selection volumes or any other sort of radius to be taken into account. If BAR or someone else wants intersecting rectangles, then fine, but I want the existing cheap system to be retained as well.

**Comment by badosu** [comment]
I'm doing my best to find solutions to the problem you presented, please don't make the assumption this is game biased. It's not about what me or you want, but what solution suits best.

**Comment by GoogleFrog** [comment]
Just to be clear, I'm not assuming game bias. I'm saying if different games want different things then there should be a toggle somewhere.

**Comment by badosu** [comment]
Idea is not too complicated actually, if it's performance characteristics are not desirable then it certainly won't be adopted.

A proper selection that includes scanning vertices retains midpos check for inside rectangle then scans vertices for collision with one spatial property that is acceptable (for performance or correctness) with units.

There are many ways to put some additional filters to make sure it is fast enough without increasing complexity considerably, for example a threshold check from mid pos to the aabb edge then, if inside the threshold, check for the collision.

**Comment by badosu** [comment]
If for example you don't care about correctness as long as the selection works 'good enough' you can interpret the selection/collision volume as a suitable 2d shape, for example an ellipsoid for spheres, ellipsoids and cylinders and it should work fast and most often well enough.

**Comment by sprunk** [comment]
I've seen some screenshots on discord:

![image](https://user-images.githubusercontent.com/2573076/223608987-0744b972-335d-4189-86ee-9ee7c02684a1.png)

If the above (tiny drag on the floating spiky ball on the left) doesn't select the unit, then the feature is useless. On the other hand just doing "if drag is empty then try point-click" should work.

![image](https://user-images.githubusercontent.com/2573076/223609164-e9630377-cce8-43ad-8806-6588c07d6803.png)
![image](https://user-images.githubusercontent.com/2573076/223609188-0b8f4b92-a87c-4b3a-b94a-a9c887f65a86.png)

If the above (drag on the edge of the gray sphere but not the actual volume) selects the unit then the feature is bad. The drag doesn't intersect any edge of an actual selection volume so shouldn't select the unit. The "if drag is empty then try point-click" would have desirable behaviour here as well.

---

After some pondering I think dragging shouldn't work on the selection volume, unless there was a separate selection volume for dragging. This is because from a UI design PoV it makes much more sense for the volume to be smaller for the purpose of dragging than for point-clicks, because that prevents selections containing more units than expected and generally lets players have more precision (it's easier to append to a selections than to deselect multiple units). The current behaviour achieves that fairly well (midpoint being a zero-size volume, effectively) and having a separate selection volume for dragging would let UI devs replicate that.

Consider SC2. Hovering over this point (cursor doesn't render in screenshot but I marked it manually) will select the nexus:
![image](https://user-images.githubusercontent.com/2573076/223611247-96e784c8-086d-4ce0-aac6-ef056d8fb6f7.png)

While dragging doesn't, its the selection volume covers only the core of the structure:
![image](https://user-images.githubusercontent.com/2573076/223611393-933c481e-8a84-4fb7-9e3c-8aeb32412e7d.png)

If the nexus used its point-click volume for dragging (and if you could select buildings alongside units in SC2, anyway) then it would notoriously end up being selected with the probes.

**Comment by badosu** [comment]
Makes sense! I'll look into OBB (oriented bounding box) intersection with selection frustum planes.
